src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,105 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        this.array = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size >= array.length) {
+            grow(size);
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index != size) {
+            checkIndex(index);
+        }
+        if (size == array.length) {
+            grow(size);
+        }
+        System.arraycopy(array, index, array, index + 1, size - index);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        int newSize = size - 1;
+        T element = array[index];
+        System.arraycopy(array, index + 1, array, index, newSize - index);
+        size--;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getIndexOf(element);
+        if (index == -1) {
+            throw new NoSuchElementException("Couldn't find element '" + element + "'!");
+        }
+        return remove(index);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public int getIndexOf(T element) {
+        for (int i = 0; i < size; i++) {
+            if ((array[i] == element) || array[i] != null && array[i].equals(element)) {
```

Comment:
```suggestion
            if (array[i] == element || array[i] != null && array[i].equals(element)) {
```
&& is evaluated before ||